rs/nervous_system/agent/src/mock.rs

```diff
@@ -0,0 +1,133 @@
+//! Ideally, we would use mockall to generate MockCallCanisters, but the
+//! CallCanisters trait is a pretty sophisticated trait. As a result, trying to
+//! make mockall generate a MockCallCanisters is maybe impossible. Therefore, we
+//! have to hand-craft this MockCallCanisters. See ./mock/tests.rs for examples.
+
+use super::*;
+use candid::Encode;
+use std::{cell::RefCell, collections::VecDeque};
+
+pub struct MockCallCanisters {
+    remaining_expected_calls: RefCell<VecDeque<ExpectedCall>>,
+}
+
+impl MockCallCanisters {
+    pub fn new() -> Self {
+        Self {
+            remaining_expected_calls: RefCell::new(VecDeque::new()),
+        }
+    }
+
+    pub fn expect_call<R: Request>(
+        &self,
+        canister_id: Principal,
+        request: R,
+        result: Result<R::Response, MockCallCanistersError>,
+    ) {
+        let request = request.payload().unwrap();
+        let result = result.map(|response| Encode!(&response).unwrap());
+
+        self.remaining_expected_calls
+            .borrow_mut()
+            .push_back(ExpectedCall {
+                canister_id,
+                request,
+                result,
+            });
+    }
+}
+
+impl Default for MockCallCanisters {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl crate::sealed::Sealed for MockCallCanisters {}
+
+impl Drop for MockCallCanisters {
+    fn drop(&mut self) {
+        pretty_assertions::assert_eq!(
+            &*self.remaining_expected_calls.borrow(),
+            &VecDeque::new(),
+            "Some expected calls were left over.",
+        );
+    }
+}
+
+impl CallCanisters for MockCallCanisters {
+    type Error = MockCallCanistersError;
+
+    fn call<R: Request>(
```

Comment:
Nit, you could use the async syntax here

```suggestion
    async fn call<R: Request>(...) -> Result<...>
```